[dev-menu] Update JS inspector URL

gabrieldonadel opened the pull request:
# Why

Closes ENG-10881

# How

Adds back `createJsInspectorMiddleware` from #25671 to the CLI and update JS inspector URL on Expo GO and expo-dev-menu to use `/_expo/debugger`

# Test Plan

Run Expo Go unversioned and dev-client through bare expo and test opening the debugger through the dev menu on Android and iOS

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

gabrieldonadel committed:
[dev-menu] Update JS inspector URL

gabrieldonadel committed:
[go] Update JS inspector URL

byCedric commented on android/expoview/src/main/java/versioned/host/exp/exponent/modules/internal/DevMenuModule.kt:
```diff
@@ -248,7 +248,7 @@ class DevMenuModule(reactContext: ReactApplicationContext, val experiencePropert
       val devSupportManager = getDevSupportManager()
       devSupportManager?.devSettings?.packagerConnectionSettings?.inspectorServerHost?.let {
-        val url = "http://$it/inspector?applicationId=${reactApplicationContext.packageName}"
+        val url = "http://$it/_expo/inspector?applicationId=${reactApplicationContext.packageName}"
```
```suggestion
        val url = "http://$it/_expo/debugger?applicationId=${reactApplicationContext.packageName}"
```

The "new" name that we use in `@react-native/dev-middleware` is "debugger". Right now, it uses the endpoint `/open-debugger`, so let's keep that somewhat similar.

byCedric commented on packages/expo-dev-menu/android/src/main/java/expo/modules/devmenu/api/DevMenuMetroClient.kt:
```diff
@@ -10,7 +10,7 @@ class DevMenuMetroClient {
 
   suspend fun openJSInspector(metroHost: String, applicationId: String) {
-    val url = Uri.parse("$metroHost/inspector")
+    val url = Uri.parse("$metroHost/_expo/inspector")
```
```suggestion
    val url = Uri.parse("$metroHost/_expo/debugger")
```

byCedric commented on packages/expo-dev-menu/ios/DevMenuDevOptionsDelegate.swift:
```diff
@@ -39,7 +39,7 @@ class DevMenuDevOptionsDelegate {
     let port = bundleURL.port ?? Int(RCT_METRO_PORT)
     let host = bundleURL.host ?? "localhost"
-    let openURL = "http://\(host):\(port)/inspector?applicationId=\(Bundle.main.bundleIdentifier ?? "")"
+    let openURL = "http://\(host):\(port)/_expo/inspector?applicationId=\(Bundle.main.bundleIdentifier ?? "")"
```
```suggestion
    let openURL = "http://\(host):\(port)/_expo/debugger?applicationId=\(Bundle.main.bundleIdentifier ?? "")"
```

byCedric committed:
fix(cli): add back legacy inspector js middleware for dev menu (#25883)

# Why

Stacked on top of #25881
Correction of #25671

This was in place for the dev menu, and I deleted it a bit too early in
https://github.com/expo/expo/pull/25671.

We can only delete this once we fully swapped over to
`@react-native/dev-middleware`, which is currently only used for the
inspector proxy infrastructure.

# How

- Added back `createJsInspectorMiddleware` from #25671

# Test Plan

- `$ bun create expo ./test-devmenu-debugger --template blank@50`
- `$ cd ./test-devmenu-debugger`
- `$ bun expo install expo-dev-client`
- `$ bun expo start`
- Press `m` in terminal
- Press "Inspect JS" in the menu within the app

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg:
https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style
Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build
(eg: updated a module plugin).

gabrieldonadel committed:
Apply suggestions from code review

Co-authored-by: Cedric van Putten <me@bycedric.com>

gabrieldonadel committed:
[go][iOS] Update JS inspector URL

gabrieldonadel committed:
Add changelog entry

brentvatne reviewed (approved):
let's merge rather than waiting for ci, since we don't want to block release for longer than necessary here and the changes look pretty scoped and well understood
